### DISCUSSION TITLE
Replace the current HA integration with this one?

### DISCUSSION THREAD

**Discussion by fucm** [pail23/stiebel_eltron_isg_component#149] [General]
Hi Paul

Would you be interested to replace the current HA stiebel eltron integration with your one? Essentially, I did implement the same thing, but can't maintain it anymore, because the modbus module can't run at the same time as the energe/SMA module on the ISG.

**Comment by Padanian** [comment]
@fucm @pail23 Hi everybody, I am from Stiebel Eltron and would like to talk to anyone of you for actively integrate heat pumps and ventilation units into home assistant.
Please contact me at alessandro.mandelli@stiebel-eltron.it

**Reply by mbeckerkassel** [reply]
Hi Alessandro. If somebody is interested i can provide the objects.

**Reply by Padanian** [reply]
I've been in contact with @pail23 for the recent months. Unfortunately, there is very little feedback on my questions from Stiebel Eltron Germany concerning the modbus issues we're facing. The modbus API is far from being perfect adherent to specs. May I forward you the questions?

**Reply by mbeckerkassel** [reply]
Yes. I would like to participate.

**Comment by fucm** [comment]
@Padanian I did respond via mail.

**Reply by Padanian** [reply]
Thanks, will get it from there

**Comment by pail23** [comment]
@fucm At the moment I don't have the capacity to polish my custom integration to the level it would meet the quality standards for being integrated into homeassistant.
In case somebody would support me (e.g. help to increase the test coverage) I would be happy to give it a try to get your integration replaced.

**Reply by fucm** [reply]
@pail23 Could you contact Alessandro (see mail above) via mail? He is asking for a live call to discuss the options for official manufacturer support - I guess this would help with he capacity problem (which I have as well).

**Reply by Padanian** [reply]
We are in contact, thanks @fucm . I'll set up a call with you both.

**Reply by ThyMYthOS** [reply]
Since this undertaking had no progress in the last year, I'd like to offer some support. My plan is to tackle the following points in order:

1. HA stiebel eltron: Remove the dependency on the HA modbus integration and directly use pymodbus
2. HA stiebel eltron: Get bronze status
3. Split the functionality of this custom integration into a python lib (python-stiebel-eltron) and the integration itself
4. (future) Support direct CAN connection instead of ISG

**Reply by mbeckerkassel** [reply]
CAN should be avoided. There is no approved documentation for this from Stiebel. The future displays and the ISG will have at least Modbus.

**Reply by fucm** [reply]
@Padanian Is there any update directly from Stiebel, w.r.t. home automation integration? And any news on dual use of ModBus and EMI on an ISG?

**Reply by Padanian** [reply]
Not that I know of.

**Comment by mbeckerkassel** [comment]
I call you on Monday with an Update

Alessandro Mandelli ***@***.***> schrieb am Fr. 24. Jan. 2025
um 20:40:

**Comment by c0ldtech** [comment]
Are there any news on this? I know this has been some time, but I found this thread just now. I'm specifically interested in getting support for cooling configuration for my LWZ/THZ model. I am currently controlling this through scripts in iobroker but would like to migrate to homeassistant. Unfortunately, so far the necessary registers do not seem to be supported by pystiebeleltron and the stiebel eltron integration. The registers are 48001 (enable cooling HK1) and 48002 (cooling temperature HK1). They correspond to the ones in my screenshot.
Support for "Kühlsystem" would be nice too, but I think that one is not exposed through modbus.

<img width="1132" height="704" alt="image" src="https://github.com/user-attachments/assets/05117445-d28a-479a-89fc-8b6af63b9b8a" />

**Reply by ThyMYthOS** [reply]
There is some movement, but it's slow. Merging PR to home assistant takes quite some time.

**Comment by Optic00** [comment]
Two recent data points that may be useful context here, since this thread has touched on how closely the Modbus interface follows its own specification.

A WPM 3i disagrees with the documented HK 1 setpoint assignment. The manual gives address 509 to the WPM 3i and 510 to WPMsystem and WPM 3, in the current revision and the previous one alike, while the machine leaves 509 unavailable and answers on 510. Checked against the controller's own Servicewelt reading at the same moment, against the neighbouring pressure, DHW and source values of the same block, and against the ISG firmware's object database, which carries no object at 30509 for model `WPM_3_I`. A second WPM 3i owner reported the same symptom on a different heat pump in #602. Details in ThyMYthOS/python-stiebel-eltron#63.

Separately, #599 collects Modbus exception 2 answers from both WPM and LWZ systems for registers the documentation lists as supported, at read starts 5219, 3679 and 3707. A register-by-register map of one WPM 3i found four refused ranges while the rest of its map read fine. Details in ThyMYthOS/python-stiebel-eltron#65.

I can provide further read-only measurements if a maintainer needs a specific data point.

**Comment by Optic00** [comment]
Picking up the question in the title.

In January 2024 the answer here was capacity, with increasing test coverage named as an example of the kind of help that would change that. Some of that has moved since, so here is where it stands from my side, together with an offer.

This component currently runs 534 tests, 530 passing and 4 skipped, at 86 percent line coverage, with the config flow at 94. The core rules ask for above 95 percent overall and 100 for the config flow, so the distance is at least a measurable one now. The largest single gaps are `diagnostics.py` at 0, `coordinator.py` at 63 and `climate.py` at 77.

The core integration has meanwhile moved a good way on its own: silver, the same `pystiebeleltron` library underneath, an async coordinator, an entity base class, a config flow, DHCP discovery, reconfiguration and diagnostics. The most visible difference left is platform coverage, since core has `climate` while this component also has sensor, binary_sensor, number, select, switch and button. I would not claim from the outside that platforms are the only difference, model handling, device lifecycle and migration are not things you can read off a file listing.

I would like to help, and would rather ask than start somewhere and duplicate work already in flight. Three candidates, roughly in order of how little needs to be settled before they can begin:

1. Closing the coverage gap in this component. Worth doing here whatever happens to the question in the title.
2. The five documentation rules currently `todo` in the core quality scale: `docs-data-update`, `docs-examples`, `docs-known-limitations`, `docs-troubleshooting`, `docs-use-cases`. They live in `home-assistant.io`, so no integration code is involved.
3. `entity-translations` and `icon-translations`, also `todo` there. This component has both, which gives a starting point, though the keys would still have to be fitted to core's entity descriptions and get through review.

All three still cost review time, so none of it comes for free to you. Porting a platform costs considerably more of it, which is why I would only start on that if it is actually wanted.

@ThyMYthOS, you have carried the core side through a long run of PRs this summer. Is help welcome there at all, and is there anything already in flight that I should keep clear of?

@pail23, a question for you as well, with no expectation of a quick answer: does the state described above change how you see the question in the title, or is capacity the limiting factor whichever way the coverage goes?

**Reply by ThyMYthOS** [reply]
The limitation is clearly review capacity on HA maintainer side. We need to make small incremental PRs that use the latest style guide of HA. I have some more PRs in the pipeline (until gold level), but development stalled due to the main developer of `modbus_connection` being on holiday.

My plan would be to iron out the differences in ISG models and versions in @pail23's integration and only port the stuff into HA that is verified on HW and works for everyone.
